contracts/modules/cover/Cover.sol

```diff
@@ -0,0 +1,570 @@
+
+    stakingPool.buyCover(
+      productId,
+      coveredAmount,
+      REWARD_BPS * premiumInNXM / BASIS_PRECISION,
+      period,
+      capacityFactor
+    );
+
+    stakingPoolsForCover[covers.length].push(
+      StakingPool(address(stakingPool), uint96(coveredAmount), uint96(premiumInNXM))
+    );
+
+    return (coveredAmount, premiumInNXM);
+  }
+
+  function extendCover(
+    uint coverId,
+    uint32 period,
+    uint96 amount,
+    uint maxPrice,
+    StakingPool[] memory stakingPools
+  ) external returns (uint) {
+    require(_isApprovedOrOwner(_msgSender(), coverId), "Cover: caller is not owner nor approved");
+
+    Cover memory cover = covers[coverId];
+    (uint newCoverId, uint priceInAsset) = _createCover(
+      ERC721.ownerOf(coverId),
+      cover.productId,
+      cover.payoutAsset,
+      amount,
+      period,
+      stakingPools
+    );
+
+    // make the cover expire at current block
+    uint32 previousPeriod = covers[coverId].period;
+
+    // new period is the remaining time
+    uint32 newPeriod = previousPeriod - (uint32(block.timestamp) - cover.start);
+    uint priceAlreadyPaid = (previousPeriod - newPeriod) / previousPeriod * cover.premium;
+    covers[coverId].period = newPeriod;
+
+    if (priceInAsset > priceAlreadyPaid) {
+      // get price for already paid asset
+      uint priceToBePaid = priceInAsset - priceAlreadyPaid;
+      require(priceToBePaid <= maxPrice, "Cover: Price exceeds maxPrice");
+      retrievePayment(priceToBePaid, cover.payoutAsset);
+    }
+
+    return newCoverId;
+  }
+
+  function addAmount(
+    uint coverId,
+    uint96 amount,
+    uint maxPrice,
+    StakingPool[] memory stakingPools
+  ) external returns (uint) {
+
+    require(msg.sender == ERC721.ownerOf(coverId), "Cover: not cover owner");
+
+    (uint coverId, uint premiumInAsset) = _addAmount(coverId, amount, stakingPools);
+
+    require(premiumInAsset <= maxPrice, "Cover: Price exceeds maxPrice");
+    retrievePayment(premiumInAsset, covers[coverId].payoutAsset);
+    return coverId;
+  }
+
+  function _addAmount(
+    uint coverId,
+    uint96 amount,
+    StakingPool[] memory stakingPools
+  ) internal returns (uint newCoverId, uint premiumInAsset) {
+
+    Cover memory previousCover = covers[coverId];
+    // clone the existing cover
+    Cover memory cover = covers[coverId];
+
+    StakingPool[] storage currentStakingPools = stakingPoolsForCover[covers.length];
+
+    uint32 period = uint32(block.timestamp) - cover.start;
+    // convert to NXM amount
+    uint amountToCover = uint(amount) * 1e18 / pool().getTokenPrice(pool().assets(cover.payoutAsset));
+    uint totalPremiumInNXM = 0;
+    for (uint i = 0; i < stakingPools.length; i++) {
+      if (amountToCover == 0) {
+        break;
+      }
+
+      IStakingPool stakingPool = IStakingPool(stakingPools[i].poolAddress);
+      (uint coveredAmount, uint premiumInNXM) = buyCoverFromPool(stakingPool, cover.productId, amountToCover, period);
```

Comment:
period should be renamed to remainingPeriod and calculated as 

`remainingPeriod = originalCover.start + originalCover.period - uint32(block.timestamp)`

Reply:
this will underflow for expired covers, but we'll check that separately anyway

Reply:
`buyCoverFromPool` -> `allocateCapacityFromPool` or smth similar; the pools doesn't "buy" a cover